src/defrag.c

```diff
@@ -840,299 +914,450 @@ int defragLaterItem(dictEntry *de, unsigned long *cursor, long long endtime, int
+
+        if (dutyCycleUs < server.active_defrag_cycle_us) {
+            /* We never reduce our cycle time, that would increase overhead.  Instead, we track this
+             *  as part of the overage, and increase wait time between cycles. */
+            defrag.timeproc_overage_us = server.active_defrag_cycle_us - dutyCycleUs;
+            dutyCycleUs = server.active_defrag_cycle_us;
         }
+    }
+    return dutyCycleUs;
+}
 
-        /* This array of structures holds the parameters for all defragmentation stages. */
-        typedef struct defragStage {
-            kvstore *kvs;
-            dictScanFunction *scanfn;
-            void *privdata;
-        } defragStage;
-        defragStage defrag_stages[] = {
-            {db->keys, defragScanCallback, db},
-            {db->expires, scanCallbackCountScanned, NULL},
-            {server.pubsub_channels, defragPubsubScanCallback,
-             &(defragPubSubCtx){server.pubsub_channels, getClientPubSubChannels}},
-            {server.pubsubshard_channels, defragPubsubScanCallback,
-             &(defragPubSubCtx){server.pubsubshard_channels, getClientPubSubShardChannels}},
-        };
-        do {
-            int num_stages = sizeof(defrag_stages) / sizeof(defrag_stages[0]);
-            serverAssert(defrag_stage < num_stages);
-            defragStage *current_stage = &defrag_stages[defrag_stage];
-
-            /* before scanning the next bucket, see if we have big keys left from the previous bucket to scan */
-            if (defragLaterStep(db, slot, endtime)) {
-                quit = 1; /* time is up, we didn't finish all the work */
-                break;    /* this will exit the function and we'll continue on the next cycle */
-            }
 
-            if (!defrag_later_item_in_progress) {
-                /* Continue defragmentation from the previous stage.
-                 * If slot is -1, it means this stage starts from the first non-empty slot. */
-                if (slot == -1) slot = kvstoreGetFirstNonEmptyDictIndex(current_stage->kvs);
-                defrag_cursor = kvstoreDictScanDefrag(current_stage->kvs, slot, defrag_cursor, current_stage->scanfn,
-                                                      &defragfns, &(defragCtx){current_stage->privdata, slot});
-            }
+/* Must be called at the end of the timeProc as it records the timeproc_end_time for use in the next
+ * computeDefragCycleUs computation. */
+static int computeDelayMs(monotime intendedEndtime) {
+    defrag.timeproc_end_time = getMonotonicUs();
+    int overage = defrag.timeproc_end_time - intendedEndtime;
+    defrag.timeproc_overage_us += overage; // track over/under desired CPU
+
+    int targetCpuPercent = server.active_defrag_cpu_percent;
+    serverAssert(targetCpuPercent > 0 && targetCpuPercent < 100);
+
+    // Given the desired duty cycle, what inter-cycle delay do we need to achieve that?
+    long totalCycleTimeUs = server.active_defrag_cycle_us * 100 / targetCpuPercent;
```

Comment:
`server.active_defrag_cycle_us` is the duty-cycle time.  The only time this is altered is when there has been starvation.  In the duty-cycle computation, we are keeping track of slight overages/underages so that we can compensate in the next cycle - achieving a perfect target CPU percentage.

If, for example, the `active_defrag_cycle_us` is 250, and the target CPU percentage is 10%, the TOTAL cycle time would be 2500us (and our 250us is 10% of that).  We would need to delay 2250us - which we round down to 2ms.

Of course then, if the timer pops after 2ms, we only need 200us to achieve 10% - so we will will still run for 250us (the config and 50us will get added to "overage".  Eventually, the overage will allow us to wait 3ms rather than 2ms, maintaining the perfect desired CPU percentage, using the configured duty cycle time.

(I'll add some additional comments to clarify.)